File: mobile/src/main/java/net/mm2d/dmsexplorer/view/view/NestedScrollingWebView.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.NestedScrollingChild;
import android.support.v4.view.NestedScrollingChildHelper;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.ViewConfiguration;
import android.webkit.WebView;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class NestedScrollingWebView extends WebView implements NestedScrollingChild {
    private final NestedScrollingChildHelper mHelper;
    private final int mTouchSlop;
    private float mStartY;
    private float mPrevY;
    private final int[] mConsumed = new int[2];
    private boolean mScrolling;

    public NestedScrollingWebView(@NonNull final Context context) {
        this(context, null);
    }

    public NestedScrollingWebView(
            @NonNull final Context context,
            @Nullable final AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NestedScrollingWebView(
            @NonNull final Context context,
            @Nullable final AttributeSet attrs,
            final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mHelper = new NestedScrollingChildHelper(this);
        setNestedScrollingEnabled(true);

        final ViewConfiguration configuration = ViewConfiguration.get(getContext());
        mTouchSlop = configuration.getScaledTouchSlop();
    }

    @Override
    public boolean onTouchEvent(final MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mStartY = event.getRawY();
                break;
            case MotionEvent.ACTION_MOVE:
                if (mScrolling) {
                    final int dy = (int) (mPrevY - event.getRawY());
                    dispatchNestedPreScroll(0, dy, mConsumed, null);
                    dispatchNestedScroll(0, mConsumed[1], 0, dy - mConsumed[1], null);
                } else {
                    if (Math.abs(mStartY - event.getRawY()) > mTouchSlop) {
                        mScrolling = true;
                        startNestedScroll(ViewCompat.SCROLL_AXIS_VERTICAL);
                    }
                }
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                if (mScrolling) {
                    stopNestedScroll();
                }
                mScrolling = false;
                break;
        }
        mPrevY = event.getRawY();
        return super.onTouchEvent(event);
    }

    public void setNestedScrollingEnabled(final boolean enabled) {
        mHelper.setNestedScrollingEnabled(enabled);
    }

    public boolean isNestedScrollingEnabled() {
        return mHelper.isNestedScrollingEnabled();
    }

    public boolean startNestedScroll(final int axes) {
        return mHelper.startNestedScroll(axes);
    }

    public void stopNestedScroll() {
        mHelper.stopNestedScroll();
    }

    public boolean hasNestedScrollingParent() {
        return mHelper.hasNestedScrollingParent();
    }

    public boolean dispatchNestedScroll(
            final int dxConsumed,
            final int dyConsumed,
            final int dxUnconsumed,
            final int dyUnconsumed,
            final int[] offsetInWindow) {
        return mHelper.dispatchNestedScroll(dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed, offsetInWindow);
    }

    public boolean dispatchNestedPreScroll(
            final int dx,
            final int dy,
            final int[] consumed,
            final int[] offsetInWindow) {
        return mHelper.dispatchNestedPreScroll(dx, dy, consumed, offsetInWindow);
    }

    public boolean dispatchNestedFling(
            final float velocityX,
            final float velocityY,
            final boolean consumed) {
        return mHelper.dispatchNestedFling(velocityX, velocityY, consumed);
    }

    public boolean dispatchNestedPreFling(
            final float velocityX,
            final float velocityY) {
        return mHelper.dispatchNestedPreFling(velocityX, velocityY);
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/view/view/NestedScrollingWebView.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.view

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.view.ViewGroup
import android.webkit.WebView
import androidx.core.view.NestedScrollingChild
import androidx.core.view.NestedScrollingChildHelper
import androidx.core.view.ViewCompat
import com.google.android.material.appbar.AppBarLayout
import kotlin.math.abs

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class NestedScrollingWebView
@JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : WebView(context, attrs, defStyleAttr), NestedScrollingChild {
    private val helper: NestedScrollingChildHelper = NestedScrollingChildHelper(this)
    private val touchSlop: Int
    private val buffer = IntArray(2)
    private var startY: Float = 0.toFloat()
    private var prevY: Float = 0.toFloat()
    private var scrolling: Boolean = false
    private var appBarLayout: AppBarLayout? = null
    private var behavior: Behavior? = null

    init {
        isNestedScrollingEnabled = true
        touchSlop = ViewConfiguration.get(context).scaledTouchSlop
        setBackgroundColor(0)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        behavior =
            (layoutParams as? androidx.coordinatorlayout.widget.CoordinatorLayout.LayoutParams)?.behavior as? Behavior
        val p = parent as? ViewGroup ?: return
        for (i in 0 until p.childCount) {
            val child = p.getChildAt(i)
            if (child is AppBarLayout) {
                appBarLayout = child
                return
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> startY = event.rawY
            MotionEvent.ACTION_MOVE -> onTouchMove(event)
            MotionEvent.ACTION_UP,
            MotionEvent.ACTION_CANCEL -> onTouchEnd()
        }
        prevY = event.rawY
        return super.onTouchEvent(event)
    }

    private fun onTouchMove(event: MotionEvent) {
        if (event.pointerCount != 1) {
            return
        }
        if (scrolling) {
            val dy = (prevY - event.rawY).toInt()
            val consumed: IntArray = buffer
            dispatchNestedPreScroll(0, dy, consumed, null)
            dispatchNestedScroll(0, consumed[1], 0, dy - consumed[1], null)
            return
        }
        if (abs(startY - event.rawY) > touchSlop) {
            scrolling = true
            startNestedScroll(ViewCompat.SCROLL_AXIS_VERTICAL)
        }
    }

    private fun onTouchEnd() {
        if (scrolling) {
            stopNestedScroll()
            scrolling = false
        }
        val appBar = appBarLayout ?: return
        appBar.setExpanded(appBar.bottom > appBar.height / 2)
    }

    override fun setNestedScrollingEnabled(enabled: Boolean) {
        helper.isNestedScrollingEnabled = enabled
    }

    override fun isNestedScrollingEnabled(): Boolean = helper.isNestedScrollingEnabled

    override fun startNestedScroll(axes: Int): Boolean {
        behavior?.scrollByUser = true
        return helper.startNestedScroll(axes)
    }

    override fun stopNestedScroll() {
        behavior?.scrollByUser = false
        helper.stopNestedScroll()
    }

    override fun hasNestedScrollingParent(): Boolean = helper.hasNestedScrollingParent()

    override fun dispatchNestedScroll(
        dxConsumed: Int,
        dyConsumed: Int,
        dxUnconsumed: Int,
        dyUnconsumed: Int,
        offsetInWindow: IntArray?
    ): Boolean = helper.dispatchNestedScroll(
        dxConsumed,
        dyConsumed,
        dxUnconsumed,
        dyUnconsumed,
        offsetInWindow
    )

    override fun dispatchNestedPreScroll(
        dx: Int,
        dy: Int,
        consumed: IntArray?,
        offsetInWindow: IntArray?
    ): Boolean = helper.dispatchNestedPreScroll(dx, dy, consumed, offsetInWindow)

    override fun dispatchNestedFling(
        velocityX: Float,
        velocityY: Float,
        consumed: Boolean
    ): Boolean = helper.dispatchNestedFling(velocityX, velocityY, consumed)

    override fun dispatchNestedPreFling(velocityX: Float, velocityY: Float): Boolean =
        helper.dispatchNestedPreFling(velocityX, velocityY)

    class Behavior(
        context: Context,
        attrs: AttributeSet?
    ) : AppBarLayout.ScrollingViewBehavior(context, attrs) {
        var scrollByUser = true
        private var prevBottom = 0

        override fun onDependentViewChanged(
            parent: androidx.coordinatorlayout.widget.CoordinatorLayout,
            child: View,
            dependency: View
        ): Boolean {
            if (!scrollByUser) {
                val dy = dependency.bottom - prevBottom
                child.scrollBy(0, dy)
            }
            prevBottom = dependency.bottom
            return super.onDependentViewChanged(parent, child, dependency)
        }
    }
}
